File: ares-forge/src/main/java/dev/tigr/ares/forge/impl/modules/movement/AutoWalk.java
package dev.tigr.ares.forge.impl.modules.movement;

import baritone.api.BaritoneAPI;
import baritone.api.pathing.goals.GoalXZ;
import dev.tigr.ares.core.feature.module.Category;
import dev.tigr.ares.core.feature.module.Module;
import dev.tigr.ares.core.setting.Setting;
import dev.tigr.ares.core.setting.settings.EnumSetting;
import net.minecraft.util.math.MathHelper;

/**
 * @author Tigermouthbear
 */
@Module.Info(name = "AutoWalk", description = "Automatically walk in a direction", category = Category.MOVEMENT)
public class AutoWalk extends Module {
    private static final int border = 30000000;
    private final Setting<Mode> mode = register(new EnumSetting<>("Mode", Mode.SMART));
    private boolean pathing = false;

    @Override
    public void onEnable() {
        if(mode.getValue() == Mode.SMART) startPathing();
    }

    @Override
    public void onMotion() {
        if(mode.getValue() == Mode.BASIC) {
            if(pathing) stopPathing();
            MC.player.moveForward = 1;
        }

        if(mode.getValue() == Mode.SMART && !pathing) startPathing();
    }

    @Override
    public void onDisable() {
        if(mode.getValue() == Mode.SMART) stopPathing();
    }

    private void startPathing() {
        GoalXZ goal = null;
        switch(getDirection()) {
            case NORTH:
                goal = new GoalXZ((int) MC.player.posX, -border);
                break;
            case NORTHEAST:
                goal = new GoalXZ(border, -border);
                break;
            case EAST:
                goal = new GoalXZ(border, (int) MC.player.posZ);
                break;
            case SOUTHEAST:
                goal = new GoalXZ(border, border);
                break;
            case SOUTH:
                goal = new GoalXZ((int) MC.player.posX, border);
                break;
            case SOUTHWEST:
                goal = new GoalXZ(-border, border);
                break;
            case WEST:
                goal = new GoalXZ(-border, (int) MC.player.posZ);
                break;
            case NORTHWEST:
                goal = new GoalXZ(-border, -border);
                break;
        }

        BaritoneAPI.getProvider().getPrimaryBaritone().getCustomGoalProcess().setGoalAndPath(goal);
        pathing = true;
    }

    private void stopPathing() {
        BaritoneAPI.getProvider().getPrimaryBaritone().getPathingBehavior().cancelEverything();
        pathing = false;
    }

    private Direction getDirection() {
        int dir = MathHelper.floor((double) (MC.player.rotationYaw * 8.0F / 360.0F) + 0.5D) & 7;
        return Direction.values()[dir];
    }

    @Override
    public String getInfo() {
        return mode.getValue().name();
    }

    enum Mode {SMART, BASIC}

    enum Direction {SOUTH, SOUTHWEST, WEST, NORTHWEST, NORTH, NORTHEAST, EAST, SOUTHEAST}
}

File: src/main/kotlin/org/aresclient/ares/impl/instrument/modules/movement/AutoWalk.kt
package org.aresclient.ares.impl.instrument.modules.movement

import baritone.api.BaritoneAPI
import baritone.api.pathing.goals.GoalXZ
import net.minecraft.util.math.MathHelper
import org.aresclient.ares.api.instruments.Module
import org.aresclient.ares.impl.util.BaritoneUtil
import org.aresclient.ares.impl.util.ChatUtil

object AutoWalk: Module(Category.MOVEMENT, "AutoWalk", "Automatically walk in a direction") {
    private const val BORDER: Int = 30000000
    
    enum class Mode {
        SMART, SIMPLE
    }

    private val mode = settings.addEnum("Mode", Mode.SMART)
    private val forceDirection = settings.addBoolean("Force Direction", false)
    private val forcedDirection = settings.addEnum("Direction", Direction.NORTH).setVisibility(forceDirection::getValue)

    private lateinit var direction: Direction
    private var pathing = false

    override fun onEnable() {
        direction = if(forceDirection.value) forcedDirection.value else getDirection()

        if(mode.value == Mode.SMART) startPathing()
    }

    public override fun onMotion() {
        if(mode.value === Mode.SIMPLE) {
            if(pathing) stopPathing()
            SELF.yaw = direction.yaw
            MC.options.forwardKey.isPressed = true
        }

        if(mode.value === Mode.SMART && !pathing) startPathing()
    }

    override fun onDisable() {
        if(mode.value == Mode.SMART) stopPathing()
        else MC.options.forwardKey.isPressed = false
    }

    private fun startPathing() {
        if(!BaritoneUtil.isBaritonePresent()) {
            mode.value = Mode.SIMPLE
            ChatUtil.error("Install Baritone to use AutoWalk smart mode")
            return
        }

        BaritoneAPI.getProvider().primaryBaritone.customGoalProcess.setGoalAndPath(when(direction) {
            Direction.NORTH -> GoalXZ(SELF.x.toInt(), -BORDER)
            Direction.NORTHEAST -> GoalXZ(BORDER, -BORDER)
            Direction.EAST -> GoalXZ(BORDER, SELF.z.toInt())
            Direction.SOUTHEAST -> GoalXZ(BORDER, BORDER)
            Direction.SOUTH -> GoalXZ(SELF.x.toInt(), BORDER)
            Direction.SOUTHWEST -> GoalXZ(-BORDER, BORDER)
            Direction.WEST -> GoalXZ(-BORDER, SELF.z.toInt())
            Direction.NORTHWEST -> GoalXZ(-BORDER, -BORDER)
        })
        pathing = true
    }

    private fun stopPathing() {
        if(!BaritoneUtil.isBaritonePresent()) return
        BaritoneAPI.getProvider().primaryBaritone.pathingBehavior.cancelEverything()
        pathing = false
    }

    private fun getDirection(): Direction {
        val dir = MathHelper.floor((SELF.yaw * 8.0f / 360.0f).toDouble() + 0.5) and 7
        return Direction.entries[dir]
    }

    enum class Direction(val yaw: Float) {
        SOUTH(0f), SOUTHWEST(45f), WEST(90f), NORTHWEST(135f), NORTH(180f), NORTHEAST(225f), EAST(270f), SOUTHEAST(315f)
    }

    override fun getInfo() = mode.value.name
}
